module/zfs/vdev_removal.c

```diff
@@ -2006,13 +2075,15 @@ spa_vdev_remove_top_check(vdev_t *vd)
 		/* available space in the pool's normal class */
 		uint64_t available = dsl_dir_space_available(
 		    spa->spa_dsl_pool->dp_root_dir, NULL, 0, B_TRUE);
-		if (available <
-		    vd->vdev_stat.vs_dspace + spa_get_slop_space(spa)) {
+		/* non-allocating vdevs have already been taken out of space */
+		if (!vd->vdev_noalloc)
+			available -= vd->vdev_stat.vs_dspace;
+		if (available < spa_get_slop_space(spa)) {
 			/*
-			 * This is a normal device. There has to be enough free
-			 * space to remove the device and leave double the
-			 * "slop" space (i.e. we must leave at least 3% of the
-			 * pool free, in addition to the normal slop space).
+			 * There has to be enough free space to remove the
+			 * device and leave double the "slop" space (i.e.
+			 * we must leave at least 3% of the pool free, in
+			 * addition to the normal slop space).
 			 */
 			return (SET_ERROR(ENOSPC));
 		}
```

Comment:
Talking this over with Matt, we have com to the conclusion that this check is actually relatively meaningless and should be removed. Since we don't actually *reserve* the space associated with the *extra* slop, it doesn't really guarantee anything. Device removal is really relying on the *normal* slop space to guarantee that the (relatively small) index for the removed device can be stored persistently even in the extreme case when the pool is "run out of space" during the removal.